plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php

```diff
@@ -351,27 +393,124 @@ public function cleanup_synchronization_state() {
 	 * @param array $batch Batch details.
 	 */
 	public function process_batch( array $batch ) : void {
+		if ( empty( $batch ) ) {
+			return;
+		}
+
 		$this->order_cache_controller->temporarily_disable_orders_cache_usage();
 
-		if ( $this->custom_orders_table_is_authoritative() ) {
-			foreach ( $batch as $id ) {
-				$order = wc_get_order( $id );
-				if ( ! $order ) {
-					$this->error_logger->error( "Order $id not found during batch process, skipping." );
-					continue;
+		$custom_orders_table_is_authoritative = $this->custom_orders_table_is_authoritative();
+		$deleted_order_ids                    = $this->process_deleted_orders( $batch, $custom_orders_table_is_authoritative );
+		$batch                                = array_diff( $batch, $deleted_order_ids );
+
+		if ( ! empty( $batch ) ) {
+			if ( $custom_orders_table_is_authoritative ) {
+				foreach ( $batch as $id ) {
+					$order = wc_get_order( $id );
+					if ( ! $order ) {
+						$this->error_logger->error( "Order $id not found during batch process, skipping." );
+						continue;
+					}
+					$data_store = $order->get_data_store();
+					$data_store->backfill_post_record( $order );
 				}
-				$data_store = $order->get_data_store();
-				$data_store->backfill_post_record( $order );
+			} else {
+				$this->posts_to_cot_migrator->migrate_orders( $batch );
 			}
-		} else {
-			$this->posts_to_cot_migrator->migrate_orders( $batch );
 		}
+
 		if ( 0 === $this->get_total_pending_count() ) {
 			$this->cleanup_synchronization_state();
 			$this->order_cache_controller->maybe_restore_orders_cache_usage();
 		}
 	}
 
+	/**
+	 * Take a batch of order ids pending synchronization and process those that were deleted, ignoring the others
+	 * (which will be orders that were created or modified) and returning the ids of the orders actually processed.
+	 *
+	 * @param array $batch Array of ids of order pending synchronization.
+	 * @param bool  $custom_orders_table_is_authoritative True if the custom orders table is currently authoritative.
+	 * @return array Order ids that have been actually processed.
+	 */
+	private function process_deleted_orders( array $batch, bool $custom_orders_table_is_authoritative ): array {
+		global $wpdb;
+
+		$deleted_from_table_name = $custom_orders_table_is_authoritative ? $this->data_store::get_orders_table_name() : $wpdb->posts;
+		$data_store_for_deletion =
+			$custom_orders_table_is_authoritative ?
+			new \WC_Order_Data_Store_CPT() :
+			wc_get_container()->get( OrdersTableDataStore::class );
+
+		$order_ids_as_sql_list = '(' . implode( ',', $batch ) . ')';
```

Comment:
let's map all values in `$batch` via `absint` function as well to prevent any possible SQLi. I don't know if we can trust `$batch` in all cases.

Reply:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/c59b8cefad3ef827bde08fd79b73c35ae622ffec (in `process_batch`, which ultimately invokes `process_deleted_orders`).